docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,673 @@
+
+```js
+app.use('*', async (req, res, next) => {
+  const url = req.originalUrl
+
+  // 1. Read index.html
+  let template = fs.readFileSync(path.resolve(__dirname, 'index.html'), 'utf-8')
+
+  // 2. Apply Vite HTML transforms. This injects the Vite HMR client,
+  //    and also applies HTML transforms from Vite plugins, e.g. global
+  //    preambles from @vitejs/plugin-react
+  template = await server.transformIndexHtml(url, template)
+
+  // 3. Load the server entry. executeEntryPoint(url) automatically transforms
+  //    ESM source code to be usable in Node.js! There is no bundling
+  //    required, and provides full HMR support.
+  const { render } = await server.ssrModuleRunner.executeEntryPoint(
+    '/src/entry-server.js',
+  )
+
+  // 4. render the app HTML. This assumes entry-server.js's exported
+  //     `render` function calls appropriate framework SSR APIs,
+  //    e.g. ReactDOMServer.renderToString()
+  const appHtml = await render(url)
+
+  // 5. Inject the app-rendered HTML into the template.
+  const html = template.replace(`<!--ssr-outlet-->`, appHtml)
+
+  // 6. Send the rendered HTML back.
+  res.status(200).set({ 'Content-Type': 'text/html' }).end(html)
+})
+```
+
+:::info ssrModuleRunner vs ssrEnvironment
+
+An alternative is to expose `server.ssrEnvironment` that implements both `ModuleExecutionEnvironment` and `ModuleRunner` as in this case, they are both in the same runtime. In this case, the code above would be:
+
+```js
+const { render } = await server.ssrEnvironment.executeEntryPoint(
+  '/src/entry-server.js',
+)
+```
+
+This may be an interesting idea if we would like to move other SSR-related methods like `ssrFixStacktrace` from the server to the `ssrEnvironment`. And we could also have `server.browserEnvironment` that extends the environment with browser-specific functionality.
+
+For this proposal, `server.ssrModuleRunner` is used to highlight the separation between the Environment instance in the server and the module runner instance even if they are both available in the server for SSR.
+
+:::
+
+## Plugins and environments
+
+The Vite server has a shared plugin pipeline, but when a module is processed it is always done in the context of a given environment.
+
+### The `hotUpdate` hook
+
+- **Type:** `(ctx: HotContext) => Array<ModuleNode> | void | Promise<Array<ModuleNode> | void>`
+- **See also:** [HMR API](./api-hmr)
```

Comment:
I don't quite understand why we need to introduce a new hook for this. Couldn't we support something like:

```js
{
  name: 'my-plugin',
  handleHotUpdate: {
    environment: true,
    handle(...) { ... }
  }
}
```

`hotUpdate` doesn't quite follow the convention like `resolveId`, `load`, `renderChunk` I think (verb+noun), especially for hooks that return something. `hotUpdate` sounds like `buildStart`, `renderStart`, which are hooks for certain events that don't return anything.

Reply:
We could add `environment: true` but I think it is better to have a clean hook for the future. At some point, everyone will migrate to environments if we move forward and it is good we don't need to use the object form forever. `handleHotUpdate` as a function will be removed as it is working with mixed modules.

I agree that `hotUpdate` may not be the best name. Maybe we could use something similar to Rollup's `watchChange`? I think `modulesChange` may be a good name. We can brainstorm other ideas, `modifyHotUpdate` may be a good description of what the hook does too

Reply:
Isn't `watchChange` like `renderStart` where they are "events" and the hooks don't return anything? I'm not sure why we have to rename `handleHotUpdate`, I think it's fine to leave it as is, especially that the type signature looks very similar to before.

Reply:
We need to rename it because `handleHotUpdate` receives a context with mixed module nodes (browser and ssr), and we need to keep that for backward compatibility. It is also called only once with module nodes from both graphs. The new function looks similar but it is called once per environment on file change, and gets a context with modules that only correspond to that environment.
As far as I can see, if we don't have a new hook here, we can't properly get backward compatibility.

Does `modifyHotUpdate` sounds better to you?

Reply:
Yeah I also noticed those differences too and it doesn't feel big to me. We can use the object form if they want to opt to a more fine-grained env-specific handling. `modifyHotUpdate` sounds odd to me 😅

Reply:
But it isn't about opting in a more fine-grained env-specific handling. The idea is that is going to be the only way to handle hot updates. We don't want to keep the mixed module graph and all the compat code forever. Ideally that all that is deleted in Vite 7. Because we can't change `handleHotUpdate` for back compat, we'll need to use the object form with a flag if we don't want a new name. And that means that in the future this will be the only special hook that you can only use with a flag and object form and can't use as a normal hook. I think it is a lot easier to just have a new hook. We can keep brainstorming a name for it until we have something that clicks for you. Maybe `controlHotUpdate`, `handleHotEvent`, `modifyHotModules`, `handleHotPropagation`, `modifyHotPropagation`.

Reply:
Yeah I see your point too. I guess on paper it's kinda unfortunate we have to rename it while looking very similar. But if we do want a different name so it's easier to migrate, `handleHotModules` and `handleHotPropagation` sounds nice. It's often mistaken that `handleHotUpdate` generally handles hot updates, but actually it's a hook that filters the modules to be propagated for hmr.